google/cloud/storage/oauth2/google_credentials.cc

```diff
@@ -173,22 +173,46 @@ CreateAuthorizedUserCredentialsFromJsonContents(std::string const& contents) {
 
 StatusOr<std::shared_ptr<Credentials>>
 CreateServiceAccountCredentialsFromJsonFilePath(std::string const& path) {
+  return CreateServiceAccountCredentialsFromJsonFilePath(
+      path, {}, google::cloud::optional<std::string>());
+}
+
+StatusOr<std::shared_ptr<Credentials>>
+CreateServiceAccountCredentialsFromJsonFilePath(
+    std::string const& path, std::set<std::string> const& scopes,
+    google::cloud::optional<std::string> const& subject) {
```

Comment:
The argument type for the optional isn't quite right. Sadly optionals in can't hold references. And the optional itself probably should be a reference to const.

In general, optional arguments are typically passed by value, then they can be inspected and moved as needed. Yes, it might incur an unnecessary copy in some cases, but I don't think it will in this case, since the string was previously being copied into the info struct anyway. So, I'm thinking something like:

```cc
StatusOr<std::shared_ptr<Credentials>>
CreateServiceAccountCredentialsFromJsonFilePath(
    std::string const& path, std::set<std::string> const& scopes,
    google::cloud::optional<std::string> subject) {
  ...
  info->subject = std::move(subject);
}
```

This should have only the same number of copies as before (and we're talking about small email address strings, so copies are likely not a big concern anyway).

Thoughts?

Reply:
And the same thing goes for `scopes`.

> it might incur an unnecessary copy in some cases, but I don't think it will in this case

It will if we take the `(!info)` return, but we shouldn't worry about that.